schema/ref.go

```diff
@@ -92,10 +106,27 @@ func (r *RefSchema) ApplyScope(scope Scope) {
 	r.referencedObjectCache = referencedObject
 }
 
+func (r *RefSchema) ValidateReferences() error {
+	if r.referencedObjectCache != nil {
+		return nil
+	}
+
+	return BadArgumentError{
+		Message: fmt.Sprintf(
+			"Ref object reference could not find an object with ID %q in namespace %q",
+			r.IDValue,
+			r.ObjectNamespace,
+		),
+	}
+}
+
 func (r *RefSchema) Unserialize(data any) (any, error) {
 	if r.referencedObjectCache == nil {
 		panic(BadArgumentError{
-			Message: "Unserialize called before ApplyScope. Did you add your RefType to a scope?",
+			Message: fmt.Sprintf(
+				"Unserialize called before ApplyScope. Did you add your RefType to a scope with the namespace %q?",
+				r.ObjectNamespace,
+			),
```

Comment:
If you want to enshrine wisdom like "`ApplyScope()` must be called before `Unserialize()`", do it in a test which can execute the code and apply the assertion from the top of the calltree -- don't bury requirements like this deep in the code at the bottom of the call tree...it's a violation of modularity which will place unwonted constraints on refactoring and future code evolution.

I'm glad that you've been helped by things like this in the past.  But, perhaps you've never worked on code which has undergone enough evolution that these "helpful tidbits" actually serve to mislead new developers.  The problem is that these bits of knowledge rot, and there is little mechanism to keep them maintained, because the parts of the code that they refer to don't refer back to them.  [And, that is _not_ an invitation to go sprinkling complementary references about the code!]  It is far better to let each piece of code stand on its own and to relegate global knowledge to global scopes (like design specifications and test implementations).

Reply:
In this case, the design for applying scopes is ingrained in the generic type interface. So it's not internal to this type. The user of the schema needs to know this, so it's information needed outside of this repository. So it should be stable, and if refactored there would be plenty of opportunity to update these error messages.

I updated the error messages to make the helpful tidbit a little more generic.